Use RBush spatial index if module is loaded

nathanntg opened the pull request:
I have been using the Werewolf package for a project, and noticed that there are some performance bottlenecks when doing a large number of lookups. Specifically, I was doing about 4,000 lookups across all counties in the US (~3,000 polygons).

In order to improve performance, I integrated a spatial index into Wherewolf. With the help of a spatial index, rather than searching through all features in the layer, Wherewolf can query for those features where the bounding box intersects with the point (the spatial index uses a tree structure to quickly identify such intersections) and only evaluate whether the query point is within the shorter list of features. (There is a [good primer on spatial indices](https://medium.com/@agafonkin/a-dive-into-spatial-search-algorithms-ebd0c5e39d2a) that inspired this change.)

I wanted to potentially contribute this change back to the community as I think it might be more widely useful. Specifically, these changes do the following:

If the [rbush](https://github.com/mourner/rbush) module is installed or loaded before Wherewolf, it will use rbush to make a spatial index of each layer and more quickly search the layer (allowing it to only search those bounding boxes that intersect with the search point).

If rbush is not available, it will continue to perform as is.

For my example, this produced substantial improvements in querying and a slight performance hit during initialization (tests run in Safari Technology Preview):

|Step|w/o rbush|w/ rbush|
|----|---:|---:|
|Initialization (loading features)|30ms|61ms|
|Running 3,000 lookups|1,016ms|67ms|

Notes:

* All tests continue to pass. 
* There may be issues associated with the antimeridian, but I am not sure.
* I added rbush to the dependencies, but that may not make sense since it is optional.

nathanntg committed:
Optionally use rbush spatial index

If the `rbush` (https://github.com/mourner/rbush) module is installed or loaded before Wherewolf, it will use rbush to make a spatial index of each layer and more quickly search the layer (allowing it to only search those bounding boxes that intersect with the search point).

veltman commented:
Thanks for the PR, I think adding RBush or an equivalent index makes sense, but it may be a while before I can review this in detail.  It might also be worth tuning `maxEntries` based on the size of the input (i.e. different balance between insertion / search speed for 10 features vs. 10,000). I haven't touched this lib in years so while I'm at it I also need to update the module bundling and such.

nathanntg commented:
No worries. I know this project has not had a lot of changes recently, but I figured this might be useful to others using it so I should at least mention it here.

I think tuning `maxEntries` automatically would be hard, since it will be dependent on the expected query volume. But I think an easy and viable option would be to expose some options when instantiating `Wherewolf` (e.g., options about whether or not to use indexing at all and, if so, parameters to pass on to the index). I can add that, but I figured it would be good to run by you before adding since it would require changing the API (adding an optional `options` argument to the constructor).

veltman commented:
In terms of tuning, I was just thinking in terms of minimizing the initialization hit in the extreme case: in a browser, on a slow phone, with a big feature set. I'd be OK with losing a bit of speed for people doing huge query volumes server-side in return. Might not be worth the trouble to tune in any case.

veltman commented:
I'm thinking that I'll end up rebuilding this to be ES6-friendly, so that bundling the relevant portions of TopoJSON directly wouldn't be prohibitive and we'd stop silently detecting it. Given that I think we'd want to make the indexing an option rather than something that's silently decided by whether RB is defined.  e.g.

```js
var wolf = new Wherewolf(..., { index: true });
```

What do you think?

nathanntg commented:
That makes sense, and I think it is good to provide that sort of API control regardless. I can try to refactor the pull request to support such a setting.

veltman commented:
I'm thinking that we could merge this into a placeholder `v1` branch and then I can refactor from there before merging back into master. Sound OK?

nathanntg committed:
Turn indexing into an option

Add an optioanal options parameters when instantiating `Wherewolf`, where indexing can be explicitly enabled. To use the spatial index, now RBush needs to be loaded and the `index` parameter must be set to true:

```
var ww = new Wherewolf({index: true});
```

Added a separate section to the  test file to explicitly test the Topojson data with indexing enabled.

In addition, I changed the spatial index parameter to optimize insertion, rather than searching.

nathanntg commented:
Yeah, whatever works best for your plans. I did go ahead and commit two changes based on our discussions:

1. Added an optioanal options parameters when instantiating `Wherewolf`, where indexing can be explicitly enabled. To use the spatial index, now RBush needs to be loaded and the `index` parameter must be set to true:

   ```
   var ww = new Wherewolf({index: true});
   ```

   I also added some new tests to ensure both the w/ index and the w/o index case are covered by the test file.

2. Per your suggestion, I adjusted RBush to prioritize faster insertion, to minimize load in cases with a low number of queries.

I did not change how I load the "rbush" change, since I am less familiar with the new ES6 module functionality and figure that you may have a specific way that you are hoping to implement it.

Let me know if there are any other changes I should make!

veltman commented:
Merging this into the `v1` branch and I'll work on the module bits from there. Thanks!
